### ISSUE TITLE
25 epochs yields sequence of <PAD> at generation 

### ISSUE THREAD

**Issue by yonatanbitton** [pbloem/language-models#4]
Hello! Thanks for the repo.
I want to train a good enough LM on the Hebrew language.
I will later want to use the model in order to get the probability of a given word in a specific context. 

I've trained your second model - 'words.py' first with 5 epochs. It seems ok, then I've taken a larger dataset (650MB of Hebrew Wikipedia) and it took around 2 days on Azure VM. 

Sadly, the output now at the generation is <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD> <PAD>
....

I'm trying to understand what went wrong.
The output model is ~65MB.
i2w & w21 are at length 10K (because of the limit). 
The real vocab size is more towards 1.2 million words. 
I suspect that this big gap is what causes bad performance.

What is good practice in order to train good LM?
Epochs? Dataset size? Words limit? Which model do you recommend out of the three suggested?

Thanks a lot :)

**Comment by pbloem** [comment]
Hi Yonatan,

Looks like Hebrew is a language with a pretty high morphological structure (lot of grammatical structure inside the words). That explains the high vocabulary. It also means that if you cut down the vocabulary, you will get a  lot of pad tokens, which is probably why the models mode collapses to just predicting the pad token.

I don't expect you'll get very far with a word-level model. You need to model the structure inside the words. The character level model may help, but it's _very_ slow. I recommend using a character-level transformer instead. It should work just as well, and train much faster. See [my blog post on the subject](http://www.peterbloem.nl/blog/transformers) for an explanation and some code samples (in pytorch rather than keras).

**Comment by yonatanbitton** [comment]
Thank you. That sounds true. Thanks for the blog post - I will read it. 
I did consider char-level LMs for the reasons you said, and i'm convinced that it will work better. 
My problem with it is the computing the probability of a word given a context. 
For example P(red | This car is).

In a word level model, it's easy: 
You feed "The car is" into the model, it predicts scores for each word in your vocab (let's say 10K dimension vector - one for each word), you normalize those scores to a probability distribution, and then you use the word-to-index in order to extract the probability for the word you want. 

I'm not sure how it's supposed to work in a character-level model, because it samples a character each time instead of word. Then I believe that multiplying the probabilities of the sampled characters may be problematic. 

Thanks a lot for your comment!

**Comment by pbloem** [comment]
You can multiply the probabilities of the characters, so long as you condition on everything before the character:
```
p(r, e, d|'this car is ') = p(r | 'this car is ') p(e|'this car is r')p(d|'this car is re')
```
This follows from the chain rule of probability. If work the same in the LSTM and the transformer model (they both model conditional probability in the same way).

**Comment by yonatanbitton** [comment]
Sounds good, it worth a try. Thanks :)
